kernel/cClosure.ml

```diff
@@ -450,6 +298,98 @@ let inject c = injectu c Univ.Instance.empty
 
 let mk_irrelevant = { mark = Cstr; term = FIrrelevant }
 
+let is_irrelevant info r =
+  info.i_cache.i_mode == Conversion &&
+  match r with
+  | Sorts.Irrelevant -> true
+  | Sorts.RelevanceVar q -> not (info.i_cache.i_sigma.qvar_relevant q)
+  | Sorts.Relevant -> false
+
+(************************************************************************)
+
+type table_val = (fconstr, Empty.t) constant_def
+
+module Table : sig
+  type t
+  val create : unit -> t
+  val lookup : clos_infos -> t -> table_key -> table_val
+end = struct
+  module Table = Hashtbl.Make(struct
+    type t = table_key
+    let equal = eq_table_key (eq_pair eq_constant_key Univ.Instance.equal)
+    let hash = hash_table_key (fun (c, _) -> Constant.UserOrd.hash c)
+  end)
+
+  type t = table_val Table.t
+
+  let create () = Table.create 17
+
+  exception Irrelevant
+
+  let shortcut_irrelevant info r =
+    if is_irrelevant info r then raise Irrelevant
+
+  let assoc_defined d =
+    match d with
+    | NamedDecl.LocalDef (_, c, _) -> inject c
+    | NamedDecl.LocalAssum (_, _) -> raise Not_found
+
+  let constant_value_in u = function
+    | Def b -> injectu b u
+    | OpaqueDef _ -> raise (NotEvaluableConst Opaque)
+    | Undef _ -> raise (NotEvaluableConst NoBody)
+    | Primitive p -> raise (NotEvaluableConst (IsPrimitive (u,p)))
+
+  let value_of info ref =
+    try
+      let env = info.i_cache.i_env in
+      match ref with
+      | RelKey n ->
+          let i = n - 1 in
```

Comment:
your `match` indentation looks a bit wrong, for instance here it should be
```suggestion
         let i = n - 1 in
```
ie
~~~
| pattern ->
  (* +2 spaces from the | position *)
  branch
~~~

Reply:
Indeed I miscounted in my suggestion, I did mean +2.

We generally do +2 but aren't always consistent. Since there are hunks of whitespace-only changes I'm noticing.

I think +4 would get pretty heavy once you have a couple nested matches.

Reply:
Yeah, that is why I avoid nesting matches as much as possible: if you need 2 or maybe 3 nested matches, you ought to have defined auxiliary functions IMO. In any case, I can do +2 if you prefer, but I honestly think this is worst.